File: leetcode/Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *	 int val;
 *	 ListNode *next;
 *	 ListNode(int x) : val(x), next(NULL) {}
 * };
 */

#include <unordered_set>

class Solution {
public:
	//typedef 
	typedef ListNode* iterator; 

	//methods
	bool hasCycle(ListNode *head) {
		if( head == NULL ){
			return false;
		}

		unordered_set< iterator > set;
		while( head != NULL ){
			if( set.find(head) != set.end() ){
				return true;
			}else{
				set.insert(head);
				head = head->next;
			}
		}

		return false;
	}
};

File: LeetCode-Java/Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *	 int val;
 *	 ListNode next;
 *	 ListNode(int x) {
 *		 val = x;
 *		 next = null;
 *	 }
 * }
 */
public class Solution {
	public boolean hasCycle(ListNode head) {
		boolean ret = false; 

		if( head == null ){
			return false; 
		}
		ListNode fst = head; 
		ListNode slo = new ListNode(0);
		slo.next = head; 

		while( fst.next != null && fst.next.next != null ){
			slo = slo.next;
			fst = fst.next.next;
			if( slo == fst ){
				return true; 
			}
		}

		return false; 
	}
}
